File: backend/app/models/answer.py
from datetime import datetime, timezone
import uuid

from sqlalchemy import UUID, String, Boolean, DateTime, ForeignKey, Numeric
from sqlalchemy.orm import Mapped, mapped_column, relationship

from dependencies.postgresql_db import Base
from models import *


def utcnow():
    return datetime.now(timezone.utc)



class Answer(Base):
    __tablename__ = "answers"
    # Columns
    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, nullable=False, index=True, primary_key=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utcnow, onupdate=utcnow)
    answer_text: Mapped[str] = mapped_column(String, nullable=True)
    has_buzzed: Mapped[bool] = mapped_column(Boolean, nullable=True, default=False)
    timestamp: Mapped[float] = mapped_column(Numeric(6, 3), nullable=True)
    # Foreign Keys
    player_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"), nullable=False)
    match_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("matches.id"), nullable=False)
    question_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("questions.id"), nullable=False)

File: packages/db/src/schema/answer.ts
import { pgTable, uuid, varchar, boolean, timestamp, numeric, index } from 'drizzle-orm/pg-core'
import { users } from './user'
import { matches } from './match'
import { questions } from './question'

/**
 * Answers table — v4 schema (unchanged from v3).
 * Stores each player's answer per question per match.
 */
export const answers = pgTable(
  'answers',
  {
    id: uuid('id').defaultRandom().primaryKey(),
    answerText: varchar('answer_text'),
    hasBuzzed: boolean('has_buzzed').default(false),
    timestamp: numeric('timestamp', { precision: 16, scale: 3 }),
    isDeleted: boolean('is_deleted').default(false),
    playerId: uuid('player_id')
      .notNull()
      .references(() => users.id),
    matchId: uuid('match_id')
      .notNull()
      .references(() => matches.id),
    questionId: uuid('question_id')
      .notNull()
      .references(() => questions.id),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow(),
    updatedAt: timestamp('updated_at', { withTimezone: true }).defaultNow(),
  },
  (t) => [
    index('idx_answers_player_id').on(t.playerId),
    index('idx_answers_match_id').on(t.matchId),
    index('idx_answers_question_id').on(t.questionId),
  ],
)
